File: edition4/js/ch03/3-1.js
var TripleStack = function() {
  this.STACK_SIZE = 100;
  this.buffer = new Array(this.STACK_SIZE * 3);
  this.topOfStacks = [0,0,0];
}

TripleStack.prototype.computeTopIndex = function(stackNum) {
  return (stackNum * this.STACK_SIZE) + this.topOfStacks[stackNum];
};

TripleStack.prototype.push = function(stackNum, value) {
  this.buffer[this.computeTopIndex(stackNum) + 1] = value;
  this.topOfStacks[stackNum] += 1;
}

TripleStack.prototype.pop = function(stackNum) {
  var value = this.buffer[this.computeTopIndex(stackNum)];
  this.topOfStacks[stackNum] -= 1;
  return value;
}

TripleStack.prototype.peek = function(stackNum) {
  return this.buffer[this.computeTopIndex(stackNum)];
}

// tests
var assert = require('assert');

describe('TripleStack', function() {

  describe('computeTopIndex', function() {

    it('should calculate correctly', function() {
      var s = new TripleStack();
      assert.equal(s.computeTopIndex(0), 0);
      s.topOfStacks[0] = 5;
      assert.equal(s.computeTopIndex(0), 5);
    });

  });

  describe('push', function() {

    it('should insert correctly', function() {
      var s = new TripleStack();
      s.push(0, 1);
      s.push(0, 2);
      assert.equal(s.peek(0), 2);
      s.push(0, 3);
      assert.equal(s.peek(0), 3);
    });

  });

  describe('pop', function() {

    it('should pop correctly', function() {
      var s = new TripleStack();
      s.push(0, 1);
      s.push(0, 2);
      s.push(0, 3);
      assert.equal(s.peek(0), 3);
      assert.equal(s.pop(0), 3);
      assert.equal(s.peek(0), 2);
    });

  });

  describe('peek', function() {

    it('should peek correctly', function() {
      var s = new TripleStack();
      s.push(0, 1);
      assert.equal(s.peek(0), 1);
      s.push(0, 2);
      assert.equal(s.peek(0), 2);
    });

  });

});

File: edition4/c/ch03/3-1.cpp
#include <assert.h>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>

using namespace std;

const int STACK_SIZE = 300;
int buffer[STACK_SIZE * 3];
int stackPointer[] = {0, 0, 0};

void push(int stackNum, int value) {
  stackPointer[stackNum]++;
  buffer[stackNum * STACK_SIZE + stackPointer[stackNum]] = value;
}

int pop(int stackNum) {
  int num = buffer[stackNum * STACK_SIZE + stackPointer[stackNum]];
  // optional: clear the vacated spot in array
  buffer[stackNum * STACK_SIZE + stackPointer[stackNum]] = 0;
  stackPointer[stackNum]--;

  return num;
}

int peek(int stackNum) {
  return buffer[stackNum * STACK_SIZE + stackPointer[stackNum]];
}

bool isEmpty(int stackNum) {
  return stackPointer[stackNum] == 0;
}

int main() {
  // tests
  assert(isEmpty(0));
  assert(isEmpty(1));
  assert(isEmpty(2));

  push(0, 1);
  push(0, 2);
  push(0, 3);

  assert(isEmpty(0) == false);

  push(1, 1);
  push(1, 2);
  push(1, 3);

  assert(isEmpty(1) == false);

  push(2, 1);
  push(2, 2);
  push(2, 3);

  assert(isEmpty(2) == false);

  assert(peek(0) == 3);
  assert(peek(1) == 3);
  assert(peek(2) == 3);

  push(0, 4);
  push(1, 4);
  push(2, 4);

  assert(peek(0) == 4);
  assert(peek(1) == 4);
  assert(peek(2) == 4);

  assert(pop(0) == 4);
  assert(pop(1) == 4);
  assert(pop(2) == 4);

  assert(peek(0) == 3);
  assert(peek(1) == 3);
  assert(peek(2) == 3);

  return 0;
}
